### PR TITLE
Migrate View JS to Controls JS

### PR DESCRIPTION
**Pull request by robwalch**
These changes migrate controls code out of view.js into it's own module. ES2015 class inheritance replaces extendable.js.

JW7-4075

### CODE REVIEW

File: src/js/view/view.js [hunk 1]
```diff
@@ -604,134 +441,52 @@ define([
...
-            _controlsLayer.appendChild(createDisplayContainer());
 
-            _dock = new Dock(_model);
+            _logo.setup(controls.right);
 
-            _logo = new Logo(_model);
-            _logo.on(events.JWPLAYER_LOGO_CLICK, _logoClickHandler);
+            controls.enable(_api, _model, _videoLayer);
+            
+            _componentFadeListeners(controls.controlbar);
```
Comment (robwalch) [inline src/js/view/view.js]: `_componentFadeListeners` should be moved to controls. These listeners need to be added and removed by controls.

File: src/js/view/view.js [hunk 2]
```diff
@@ -460,6 +310,11 @@ define([
             _title = new Title(_model);
             _title.setup(_titleElement);
 
+            _logo = new Logo(_model);
+            _logo.setup(_playerElement);
+            _logo.on(events.JWPLAYER_LOGO_CLICK, _logoClickHandler);
+            _componentFadeListeners(_logo);
```
Comment (robwalch) [inline src/js/view/view.js]: Move `_componentFadeListeners` to controls.

File: src/js/utils/css.js [hunk 3]
```diff
@@ -101,33 +101,41 @@ define([
...
-            parseInt(hexColor.substr(4, 2), 16)
-        ];
-        if (opacity !== undefined && opacity !== 100) {
+        var hasAlpha = (opacity !== undefined && opacity !== 100);
+        if (hasAlpha) {
             style += 'a';
-            channels.push(opacity / 100);
         }
-        return style + '(' + channels.join(',') + ')';
+        if (!canvasColorContext) {
```
Comment (RikHeijdens) [inline src/js/utils/css.js]: @robwalch Out of curiosity, how does using the canvas compare performance-wise to the old style of calculating the rgba values?

Comment (robwalch) [inline src/js/utils/css.js]: It lets us use any CSS color like "red", not just hex, and it's fast enough that I'm not concerned about performance impact. There may be a little time/memory hit getting the context. After that it's pretty clean - we'd have to run tests to see which is slower and which cause more GC as a result of memory usage.

Comment (robwalch) [inline src/js/utils/css.js]: When doing timeline profiling of view setup, I've never seen a point where this code stood out. Moving out of the view and updating the util removes code that duplicates functionality and ensures that we only need this one context for all players on the page.

Comment (RikHeijdens) [inline src/js/utils/css.js]: Interesting - I didn't realize it would handle arbitrary CSS colors, I like it.

File: src/js/controller/controls-loader.js [hunk 4]
```diff
@@ -1,7 +1,7 @@
 define([
 
 ], function() {
-    let controls = null;
```
Comment (johnBartos) [inline src/js/controller/controls-loader.js]: Can keep this as `let`

File: src/js/view/controls/clickhandler.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
+define([
```
Comment (johnBartos) [inline src/js/view/controls/clickhandler.js]: I suppose we could ditch `define` once the ES6 PR gets merged

File: src/js/view/controls/components/chapters.mixin.js [hunk 6]
```diff
@@ -4,28 +4,29 @@ define([
...
 ], function(_, utils, srt) {
 
-    function Cue(time, text) {
-        this.time = time;
-        this.text = text;
-        this.el = document.createElement('div');
-        this.el.className = 'jw-cue jw-reset';
-    }
+    class Cue {
+
```
Comment (johnBartos) [inline src/js/view/controls/components/chapters.mixin.js]: No unnecessary whitespace

File: src/js/view/controls/components/chapters.mixin.js [hunk 7]
```diff
@@ -4,28 +4,29 @@ define([
...
+        constructor (time, text) {
+            this.time = time;
+            this.text = text;
+            this.el = document.createElement('div');
```
Comment (johnBartos) [inline src/js/view/controls/components/chapters.mixin.js]: Do we want to consider passing `document` as something like `context` for these components so they can be easily unit tested?

File: src/js/view/controls/components/chapters.mixin.js [hunk 8]
```diff
@@ -34,38 +35,37 @@ define([
...
             // We won't want to draw them until we have a duration
-            var duration = this._model.mediaModel.get('duration');
+            const duration = this._model.mediaModel.get('duration');
             if (!duration || duration <= 0) {
                 this._model.mediaModel.once('change:duration', this.drawCues, this);
                 return;
             }
 
-            var _this = this;
+            const self = this;
```
Comment (johnBartos) [inline src/js/view/controls/components/chapters.mixin.js]: `() =>` and don't use `sel`f

File: src/js/view/controls/components/slider.js [hunk 9]
```diff
@@ -26,11 +31,10 @@ define([
...
 
             this.tapListener = this.tap.bind(this);
+        }
 
-            this.setup();
-        },
-        setup: function() {
-            var obj = {
+        setup() {
+            const obj = {
```
Comment (johnBartos) [inline src/js/view/controls/components/slider.js]: I think we can come up with a better name than `obj`

File: src/js/view/controls/components/timeslider.js [hunk 10]
```diff
@@ -53,18 +57,22 @@ define([
...
+    }
 
     function reasonInteraction() {
         return { reason: 'interaction' };
     }
 
-    var TimeSlider = Slider.extend({
-        constructor: function(_model, _api) {
+    class TimeSlider extends Slider {
+
```
Comment (johnBartos) [inline src/js/view/controls/components/timeslider.js]: Remove newline before constructor

File: src/js/view/controls/components/volumetooltip.js [hunk 11]
```diff
@@ -1,17 +1,21 @@
...
+    'view/controls/components/slider',
     'utils/ui',
     'utils/helpers'
 ], function(Tooltip, Slider, UI) {
-    return Tooltip.extend({
-        constructor: function(_model, name, ariaText) {
-            this._model = _model;
 
+    return class VolumeTooltip extends Tooltip {
+
```
Comment (johnBartos) [inline src/js/view/controls/components/volumetooltip.js]: Newline

File: src/js/view/controls/controls.js [hunk 12]
```diff
@@ -1,31 +1,386 @@
...
+    'view/controls/rightclick',
+], function (events, states, Events, Constants, utils, _, button, ClickHandler, Controlbar, Dock,
+             DisplayContainer, RewindDisplayIcon, PlayDisplayIcon, NextDisplayIcon,
+             NextUpToolTip, RightClick) {
+
+    const ACTIVE_TIMEOUT = utils.isMobile() ? 4000 : 2000;
+    const CONTOLBAR_ONLY_HEIGHT = 44;
+
+    const isAudioMode = function(model) {
+        let playerHeight = model.get('height');
```
Comment (johnBartos) [inline src/js/view/controls/controls.js]: `const`

File: src/js/view/controls/controls.js [hunk 13]
```diff
@@ -1,31 +1,386 @@
...
+            this.rightClickMenu = null;
+            this.showing = false;
+            this.unmuteCallback = null;
 
             const element = this.context.createElement('div');
             element.className = 'jw-controls jw-reset';
             this.element = element;
 
-            this.enable();
+            const right = document.createElement('div');
```
Comment (johnBartos) [inline src/js/view/controls/controls.js]: Not thrilled about doing work in a constructor, can this be broken out easily?

File: src/js/view/controls/controls.js [hunk 14]
```diff
@@ -1,31 +1,386 @@
...
+                this.element.appendChild(displayContainer.element());
+                this.displayContainer = displayContainer;
+            }
+
+
+            // Display Click and Double Click Handling
+            const displayClickHandler = new ClickHandler(model, videoLayer, { useHover: true });
+            displayClickHandler.on({
+                click: () => {
+                    this.trigger(events.JWPLAYER_DISPLAY_CLICK);
```
Comment (johnBartos) [inline src/js/view/controls/controls.js]: Trigger last?

File: src/js/view/controls/controls.js [hunk 15]
```diff
@@ -1,31 +1,386 @@
...
+                },
+                tap: () => {
+                    this.trigger(events.JWPLAYER_DISPLAY_CLICK);
+                    const state = model.get('state');
+
+                    if (((state === states.IDLE || state === states.COMPLETE) ||
+                        (this.instreamState === states.PAUSED))) {
+                        api.play(reasonInteraction());
+                    }
+                    if (state === states.PAUSED) {
```
Comment (johnBartos) [inline src/js/view/controls/controls.js]: If we're `PAUSED` and  instreamState is `PAUSED` are we supposed to enter both `if`s?

File: src/js/view/controls/controls.js [hunk 16]
```diff
@@ -1,31 +1,386 @@
...
         }
+
+        resize(model, breakPoint) {
+            const audioMode = isAudioMode(model);
+
+            // Set timeslider flags
+            const smallPlayer = breakPoint < 2;
+            const timeSliderAboveConfig = model.get('timeSliderAbove');
+            const timeSliderAbove = !audioMode &&
+                (timeSliderAboveConfig !== false) && (timeSliderAboveConfig || smallPlayer);
```
Comment (johnBartos) [inline src/js/view/controls/controls.js]: I'd put a newline after the declarations

File: src/js/view/controls/display-container.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+define([
+    'utils/helpers',
+    'templates/display-container.html'
+], function(utils, Template) {
+
+    return class DisplayContainer {
+
```
Comment (johnBartos) [inline src/js/view/controls/display-container.js]: Remove newline

File: src/js/view/controls/dock.js [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+            // Clicks on the text overlay
+            return evt.target.parentElement.parentElement;
+        }
+
+        // Clicks on any other children
+        return evt.target.parentElement;
+    };
+
+    return class Dock {
+
```
Comment (johnBartos) [inline src/js/view/controls/dock.js]: Remove newline

File: src/js/view/controls/next-display-icon.js [hunk 19]
```diff
@@ -0,0 +1,30 @@
+define([
+    'utils/helpers',
+    'utils/ui',
+    'templates/display-icon.html'
+], function(utils, UI, Template) {
+
+    return class NextDisplayIcon {
+
```
Comment (johnBartos) [inline src/js/view/controls/next-display-icon.js]: Remove newline

File: src/js/view/controls/next-display-icon.js [hunk 20]
```diff
@@ -0,0 +1,30 @@
+define([
+    'utils/helpers',
+    'utils/ui',
+    'templates/display-icon.html'
```
Comment (johnBartos) [inline src/js/view/controls/next-display-icon.js]: It would be great if we could somehow remove template imports so we don't need the handlebars-loader when running unit tests

File: src/js/view/controls/nextuptooltip.js [hunk 21]
```diff
@@ -78,30 +82,32 @@ define([
...
             dom.toggleClass(this.container, 'jw-nextup-sticky', !!model.get('nextUpSticky'));
-        },
-        setNextUpItem: function(nextUpItem) {
-            var _this = this;
+        }
+
+        setNextUpItem(nextUpItem) {
+            var self = this;
             // Give the previous item time to complete its animation
             setTimeout(function () {
```
Comment (johnBartos) [inline src/js/view/controls/nextuptooltip.js]: `() => !`

File: src/js/view/controls/rewind-display-icon.js [hunk 22]
```diff
@@ -0,0 +1,35 @@
+define([
+    'utils/helpers',
+    'utils/ui',
+    'templates/display-icon.html'
+], function(utils, UI, Template) {
+
+    return class RewindDisplayIcon {
+
```
Comment (johnBartos) [inline src/js/view/controls/rewind-display-icon.js]: Remove newline

File: src/js/view/controls/rewind-display-icon.js [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+        constructor(model, api) {
+
+            this.el = utils.createElement(Template({
+                iconName: 'rewind',
+                ariaLabel: model.get('localization').playback
```
Comment (johnBartos) [inline src/js/view/controls/rewind-display-icon.js]: Remove newline

File: src/js/controller/controls-loader.js (hunk 4 above)
Comment (robwalch) [inline src/js/controller/controls-loader.js]: The loader is not part of jwplayer.controls.js so I moved it out of the controls dir. No babel loading there until after #1927 is merged.

File: src/js/view/controls/clickhandler.js (hunk 5 above)
Comment (robwalch) [inline src/js/view/controls/clickhandler.js]: 👍

File: src/js/view/controls/components/chapters.mixin.js (hunk 6 above)
Comment (robwalch) [inline src/js/view/controls/components/chapters.mixin.js]: Shouldn't that be an ESLint rule?

File: src/js/view/controls/components/chapters.mixin.js (hunk 7 above)
Comment (robwalch) [inline src/js/view/controls/components/chapters.mixin.js]: Good idea. It would also help use deal with #1138

File: src/js/view/controls/components/slider.js (hunk 9 above)
Comment (robwalch) [inline src/js/view/controls/components/slider.js]: I think you can come up with a better comment...

File: src/js/view/controls/controls.js (hunk 14 above)
Comment (robwalch) [inline src/js/view/controls/controls.js]: In this case no because `api.play` will result in events being dispatched that should come after the click event.

This is an example where DOM events with `preventDefault` could add functionality for developers who would like to override our default click behavior.

File: src/js/view/controls/controls.js (hunk 15 above)
Comment (robwalch) [inline src/js/view/controls/controls.js]: The second if is for toggling controls on mobile. So yes.

File: src/js/view/controls/controls.js (hunk 16 above)
Comment (robwalch) [inline src/js/view/controls/controls.js]: Why?

File: src/js/view/controls/rewind-display-icon.js (hunk 22 above)
Comment (robwalch) [inline src/js/view/controls/rewind-display-icon.js]: I like having a space above the constructor. It makes the class and constructor more readable.

File: src/js/view/controls/rewind-display-icon.js (hunk 23 above)
Comment (robwalch) [inline src/js/view/controls/rewind-display-icon.js]: Stop doing the work of a linter. This kind of feedback is counter productive.

File: src/js/view/controls/controls.js (hunk 16 above)
Comment (johnBartos) [inline src/js/view/controls/controls.js]: It's neater and you do it in other places

File: src/js/view/controls/rewind-display-icon.js (hunk 23 above)
Comment (johnBartos) [inline src/js/view/controls/rewind-display-icon.js]: There's no lint rule for htis

File: src/js/view/controls/components/chapters.mixin.js (hunk 6 above)
Comment (robwalch) [inline src/js/view/controls/components/chapters.mixin.js]: @johnBartos check out https://github.com/eslint/eslint/issues/5949

I'm happy with this as-is.

Comment (johnBartos) [inline src/js/view/controls/components/chapters.mixin.js]: Ok np

File: src/js/view/controls/next-display-icon.js (hunk 20 above)
Comment (robwalch) [inline src/js/view/controls/next-display-icon.js]: So use template strings and ditch handlebars? :)

Comment (johnBartos) [inline src/js/view/controls/next-display-icon.js]: I think we loose too much; formatting HTML just with strings is extremely tedious. I suppose the actual problem is avoiding the need for handlebars when running unit tests, which is caused by requiring these templataes

File: src/js/view/controls/controls.js [hunk 24]
```diff
@@ -1,31 +1,361 @@
...
+                    this.trigger(events.JWPLAYER_DISPLAY_CLICK);
+                    this.userActive(1000);
+                    api.play(reasonInteraction());
+                });
+                // make playDisplayIcon clickthrough on chrome for flash to avoid power safe throttle
+                if (utils.isChrome() && !utils.isMobile()) {
+                    playDisplayIcon.el.addEventListener('mousedown', () => {
+                        const provider = model.getVideo();
+                        const isFlash = (provider && provider.getName().name.indexOf('flash') === 0);
+                        if (!isFlash) {
```
Comment (johnBartos) [inline src/js/view/controls/controls.js]: `if (provider && provider.getName().name.indexOf('flash') === 0)`; saves a variable

File: src/js/view/controls/controls.js [hunk 25]
```diff
@@ -1,31 +1,361 @@
...
+                (timeSliderAboveConfig !== false) && (timeSliderAboveConfig || smallPlayer);
+            utils.toggleClass(this.playerContainer, 'jw-flag-small-player', smallPlayer);
+            utils.toggleClass(this.playerContainer, 'jw-flag-audio-player', audioMode);
+            utils.toggleClass(this.playerContainer, 'jw-flag-time-slider-above', timeSliderAbove);
+
+            model.set('audioMode', audioMode);
+        }
+
+        unmuteAutoplay(api, model) {
+            const autostartSucceeded = !model.get('autostartFailed');
```
Comment (johnBartos) [inline src/js/view/controls/controls.js]: Is `autostartFailed` initialized to `undefined` or `false`?

File: src/js/view/controls/next-display-icon.js [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+
+    return class NextDisplayIcon {
+
+        constructor(model, api) {
+            const element = utils.createElement(Template({
+                iconName: 'next',
+                ariaLabel: model.get('localization').next
+            }));
+
+            this.iconUI = new UI(element).on('click tap', function() {
```
Comment (johnBartos) [inline src/js/view/controls/next-display-icon.js]: In `dock.js` you have a setup function which is called in the constructor, here you do everything in the constructor. Not sure which is better but should we choose 1 style and stick to it?

File: src/js/view/utils/clickhandler.js [hunk 27]
```diff
@@ -0,0 +1,72 @@
+define([
+    'utils/ui',
+    'events/events',
+    'utils/backbone.events',
+    'utils/underscore'
+], function(UI, events, Events, _) {
```
Comment (johnBartos) [inline src/js/view/utils/clickhandler.js]: I think it would be better to replace the clickHandlers when needed instead of doing this `alternateClickHandler` stuff.

File: src/js/view/controls/controls.js (hunk 16 above)
Comment (robwalch) [inline src/js/view/controls/controls.js]: done

File: src/js/view/controls/components/chapters.mixin.js [hunk 28]
```diff
@@ -34,50 +35,42 @@ define([
...
+            _.each(this.cues, (cue) => {
                 cue.align(duration);
-                cue.el.addEventListener('mouseover', function () {
-                    _this.activeCue = cue;
-                });
-                cue.el.addEventListener('mouseout', function () {
-                    _this.activeCue = null;
-                });
-                _this.elementRail.appendChild(cue.el);
+                cue.el.addEventListener('mouseover', () => (this.activeCue = cue));
```
Comment (robwalch) [inline src/js/view/controls/components/chapters.mixin.js]: The linter then complains about brace-style 🤓

File: src/js/view/controls/controls.js (hunk 24 above)
Comment (robwalch) [inline src/js/view/controls/controls.js]: There's a `!` not there and the variable makes it readable.

File: src/js/view/controls/controls.js (hunk 25 above)
Comment (robwalch) [inline src/js/view/controls/controls.js]: @egreaves implemented this. It is not set by default, and later set to `undefined`. search.

File: src/js/view/utils/clickhandler.js (hunk 27 above)
Comment (robwalch) [inline src/js/view/utils/clickhandler.js]: Let's work out how ad click functionality should work with ads team, and how we can separate it from display UI. This is the current working implementation.

File: src/js/view/controls/controls.js (hunk 24 above)
Comment (johnBartos) [inline src/js/view/controls/controls.js]: 👌

File: src/js/view/utils/clickhandler.js (hunk 27 above)
Comment (johnBartos) [inline src/js/view/utils/clickhandler.js]: Sure
